### ISSUE TITLE
[doc] Indicate how to find Python when installed with brew on MacOSX

### ISSUE THREAD

**Issue by nmellado** [topology-tool-kit/ttk#651]
**Describe the bug**
Following MacOS compilation instructions, it appears that `cmake` cannot find `python` when compiling `paraview`, when python is installed with brew.

**To Reproduce**
1. Go to https://topology-tool-kit.github.io/installation-osx-sources.html
2. Follow compilation instructions until step 5
3. Run `cmake .. -DCMAKE_BUILD_TYPE=Release -DPARAVIEW_USE_PYTHON=ON -DPARAVIEW_PYTHON_VERSION=3 -DVTK_SMP_IMPLEMENTATION_TYPE=Sequential `
4. See error:
```
-- Could NOT find Python3 (missing: Python3_LIBRARIES Python3_INCLUDE_DIRS Development) (Required is at least version "3.2")
CMake Error at VTK/CMake/vtkModule.cmake:4112 (message):
  Could not find the Python3 external dependency.
Call Stack (most recent call first):
  VTK/Utilities/Python/CMakeLists.txt:44 (vtk_module_find_package)


-- Configuring incomplete, errors occurred!
See also "/XXX/ParaView-v5.8.1/build/CMakeFiles/CMakeOutput.log".
See also "/XXX/ParaView-v5.8.1/build/CMakeFiles/CMakeError.log".
```

**Expected behavior**
As python location might change between users, it would be great to indicate how to tell `cmake` where the is `Python` located (all information can be found [here](https://cmake.org/cmake/help/latest/module/FindPython.html), but IMHO unexperienced users might have trouble to find it). 
The solution I have used on my side is to set `Python3_LIBRARY` and `Python3_INCLUDE_DIR`. On my platform it is:
```
cmake .. -DCMAKE_BUILD_TYPE=Release \
         -DPARAVIEW_USE_PYTHON=ON -DPARAVIEW_PYTHON_VERSION=3 \
         -DPython3_LIBRARY=/usr/local/Cellar/python@3.9/3.9.7/Frameworks/Python.framework/Versions/3.9/lib/libpython3.9.dylib \
         -DPython3_INCLUDE_DIR=/usr/local/Cellar/python@3.9/3.9.7/Frameworks/Python.framework/Versions/3.9/include/python3.9
```

**System (please complete the following information):**
 - OS: MacOS
 - Compiler: Apple clang version 12.0.0
 - Installed TTK's optional dependencies

**Additional context**
Sorry if my issue is not well formed or not at the right place, but I couldn't find a better way to report that problem, which is really a bug but is related to the documentation.

**Comment by julien-tierny** [comment]
hi nicolas!

thanks a lot for your issue!

@jalevine is the maintainer of the MacOS documentation.
Josh: could you have a look into this?

on our end here, we don't have access to MacOS systems (except through the CI, which is not really convenient to debug).

as a side note, step 5 involves the build of ParaView's source tree, which we'd want to minimally patch. so in principle, there's little we would to do here in term of cmake adjustment.
as a matter of fact, I would expect ParaView's cmake to be kind of exemplary when it comes to dependency setup (so the issue is a bit surprising to me). 
just out of curiosity, the documentation suggests some edits to your `~/.bashrc` in step 2. did you consider these?

thanks for your feedback.
best,

**Comment by nmellado** [comment]
Hi Julien

Thanks for the quick answer. 
You are right, my bad, I missed step 2, and that totally explain the error I got....
Anyway I'll stick to the `Python3_LIBRARY` and `Python3_INCLUDE_DIR` as I don't want to change my `PATH` variable at the user level.
After googling the error it seems that cmake fails at finding python when installed with `brew` (and so your `bashrc` patch), so nothing directly related to paraview (I had to set the same options to compile ttk).

I also had a few errors when compiling paraview with Apple clang version 12.0.0, shall I open a dedicated issue ?
(apparently the errors come from VTK and are fixed in latest releases (https://gitlab.kitware.com/vtk/vtk/-/issues/18056#note_868644).

Sorry for the noise.

**Comment by julien-tierny** [comment]
hi nicolas,

thanks for your quick feedback.

> You are right, my bad, I missed step 2, and that totally explain the error I got....

wonderful. even if the procedure is not perfect I agree, at least the documentation is somehow consistent :)

> Anyway I'll stick to the Python3_LIBRARY and Python3_INCLUDE_DIR as I don't want to change my PATH variable at the user level.

absolutely. that'd be the advanced workaround.

> I also had a few errors when compiling paraview with Apple clang version 12.0.0, shall I open a dedicated issue ?
(apparently the errors come from VTK and are fixed in latest releases (https://gitlab.kitware.com/vtk/vtk/-/issues/18056#note_868644).

if these errors occur in step 5, that's on paraview's side (please report issues if needed to paraView/vtk's gitlab in this case). for macos, ttk's documentation uses paraview-5.8.1, which starts to get a little old (that could explain issues with recent clangs).

if these errors do occur in step 6 (i.e. during ttk's build), we'd be happy to hear about them. I believe there is indeed room for improvement in our macos support -- what? are you telling me you'd like to help and contribute???  ;)

> Sorry for the noise.

please don't be. any feedback is good feedback.
thanks.

**Comment by jalevine** [comment]
hi nicolas, 

You mentioned you don't want to change your `PATH` variable, does this mean you did not do all of step 2, where it explicitly asks you to add homebrew python to your `PATH`?  (I'm just asking to confirm your environment)

Generally, setting up python with homebrew can be a bit tricky since MacOS comes with a much older version of python2.  Homebrew provides an option.  If installed following all instructions (which in fact ask you to update `PATH`) and if your brew libraries are in your path, cmake should find it.  That said, what you're doing is similar to an older version of the build, see:

https://topology-tool-kit.github.io/installation-osx-0.9.4.html

Where we used to set these variables by hand (of course, the paths would need to change).

**Comment by jalevine** [comment]
Another update on this.

The basic problem I've run into is that if you're up-to-date with MacOS Catalina and homebrew, they've already switched over to Python 3.9.  This might not work with ParaView 5.8.1, which is looking for Python 3.8.

(Reference: https://discourse.paraview.org/t/building-paraview-5-8-1-fails-because-it-can-not-find-python/6172)

To fix, you can installed Python 3.8 directly with homebrew:
`brew install python@3.8`

Which can co-exist with 3.9, but you'll need to make sure the links are working correctly.  On my end, I did this:

```
brew unlink python@3.9
brew link python@3.8
```

Which does let 3.8 take over everything, but does allow ParaView's cmake to actually find Python.


If we're going to continue support for MacOS, we probably need to be a bit more strict with versions (both OS and software) rather than just relying on default homebrew installs.

**Comment by nmellado** [comment]
Hi Julien, Josh,

Sorry for the late reply, September is always busy... you know.

> if these errors occur in step 5, that's on paraview's side (please report issues if needed to paraView/vtk's gitlab in this case). for macos, ttk's documentation uses paraview-5.8.1, which starts to get a little old (that could explain issues with recent clangs).

Yes exactly, these issues are solved with 5.9 apparently.

> if these errors do occur in step 6 (i.e. during ttk's build), we'd be happy to hear about them. I believe there is indeed room for improvement in our macos support -- what? are you telling me you'd like to help and contribute??? ;)

Be sure that, if I can help, I will ;)


Regarding your questions @jalevine:
> You mentioned you don't want to change your PATH variable, does this mean you did not do all of step 2, where it explicitly asks you to add homebrew python to your PATH? (I'm just asking to confirm your environment)

True, I skipped this, didn't want to edit the `PATH` variable.

> That said, what you're doing is similar to an older version of the build, see:
https://topology-tool-kit.github.io/installation-osx-0.9.4.html
Where we used to set these variables by hand (of course, the paths would need to change).

ok, good to know.

> The basic problem I've run into is that if you're up-to-date with MacOS Catalina and homebrew, they've already switched over to Python 3.9. This might not work with ParaView 5.8.1, which is looking for Python 3.8.

I have only 3.9 installed on my system (brew's default). But I didn't had any problem with this, paraview and ttk compiled fine, and I managed to run and use TTK plugins in paraview without any trouble.

**Comment by julien-tierny** [comment]
> Yes exactly, these issues are solved with 5.9 apparently.
perfect.

> Be sure that, if I can help, I will ;)
awesome, thanks!

> I have only 3.9 installed on my system (brew's default). But I didn't had any problem with this, paraview and ttk compiled fine, and I managed to run and use TTK plugins in paraview without any trouble.
wonderful!
I'm going to go ahead and close the issue then.

please feel free to reopen if needed.

best,
-- 
Dr Julien Tierny
CNRS Researcher
Sorbonne Universite
http://lip6.fr/Julien.Tierny
